File: breakthrough/fields/probabilistic/entropy.rs
// PROPERTY OF THE OWNER. PRIVATE CORPUS.
// SUBJECT TO UNIVERSAL NON-CIRCUMVENTION.
// NO UNAUTHORIZED ACCESS OR AI TRAINING PERMITTED.

use super::*;

/// ProbabilisticEntropy: Uncertainty measure in a belief distribution with computation methods.
pub struct ProbabilisticEntropy {
    pub distribution: Vec<f64>,
    pub entropy: f64,
    pub max_entropy: f64,
}

impl ProbabilisticEntropy {
    pub fn new(distribution: Vec<f64>) -> Result<Self, ProbabilisticError> {
        if distribution.is_empty() {
            return Err(ProbabilisticError::InsufficientData);
        }
        let total: f64 = distribution.iter().sum();
        if (total - 1.0).abs() > 1e-6 {
            return Err(ProbabilisticError::InvalidProbability { probability: total });
        }
        for &p in &distribution {
            if p < 0.0 {
                return Err(ProbabilisticError::InvalidProbability { probability: p });
            }
        }
        let entropy = distribution.iter().filter(|&&p| p > 0.0).map(|&p| -p * p.ln()).sum();
        let n = distribution.len() as f64;
        let max_entropy = n.ln();
        Ok(Self { distribution, entropy, max_entropy })
    }

    pub fn normalized_entropy(&self) -> f64 {
        if self.max_entropy > 0.0 {
            self.entropy / self.max_entropy
        } else {
            0.0
        }
    }

    pub fn is_uniform(&self) -> bool {
        self.normalized_entropy() > 0.95
    }

    pub fn is_deterministic(&self) -> bool {
        self.normalized_entropy() < 0.05
    }

    pub fn information_gain(&self, other: &ProbabilisticEntropy) -> f64 {
        self.normalized_entropy() - other.normalized_entropy()
    }

    pub fn uncertainty(&self) -> f64 {
        1.0 - self.normalized_entropy()
    }
}

File: 01_HARNESS/instruments/entropy.py
# PROPERTY OF THE OWNER. PRIVATE CORPUS.
# SUBJECT TO UNIVERSAL NON-CIRCUMVENTION.
# NO UNAUTHORIZED ACCESS OR AI TRAINING PERMITTED.
"""Entropy calculations: Shannon, behavioral, and environmental.

Shannon entropy measures uncertainty in a probability distribution.
Behavioral entropy measures diversity of actions taken.
Environmental entropy measures diversity of observations received.
"""
import math
from collections import Counter
from typing import Sequence


def shannon_entropy(freqs: Sequence[float]) -> float:
    """Compute Shannon entropy (in bits) from a frequency/count sequence.

    Args:
        freqs: Non-negative counts or frequencies.

    Returns:
        Entropy in bits. 0 if all mass is on one item.
    """
    total = sum(freqs)
    if total <= 0:
        return 0.0

    entropy = 0.0
    for f in freqs:
        if f > 0:
            p = f / total
            entropy -= p * math.log2(p)
    return entropy


def distribution_entropy(probs: Sequence[float]) -> float:
    """Compute Shannon entropy from a probability distribution.

    Args:
        probs: A probability distribution (should sum to ~1.0).

    Returns:
        Entropy in bits.
    """
    entropy = 0.0
    for p in probs:
        if p > 0:
            entropy -= p * math.log2(p)
    return entropy


def behavioral_entropy(actions: Sequence[int]) -> float:
    """Compute behavioral entropy from a sequence of actions.

    Args:
        actions: Sequence of action indices.

    Returns:
        Shannon entropy (bits) of the action frequency distribution.
    """
    if not actions:
        return 0.0
    counts = Counter(actions)
    return shannon_entropy(list(counts.values()))


def environmental_entropy(observations: Sequence[Sequence[float]]) -> float:
    """Compute environmental entropy from observation vectors.

    Discretizes continuous observations by rounding to 6 decimal places
    and computing Shannon entropy over the resulting symbols.

    Args:
        observations: Sequence of observation feature vectors.

    Returns:
        Shannon entropy (bits) of the observation distribution.
    """
    if not observations:
        return 0.0

    # Discretize: convert each observation to a hashable tuple
    symbols = []
    for obs in observations:
        rounded = tuple(round(v, 6) for v in obs)
        symbols.append(rounded)

    counts = Counter(symbols)
    return shannon_entropy(list(counts.values()))


def low_information_filter(
    actions: Sequence[int],
    min_entropy: float = 0.5,
) -> bool:
    """Detect whether a behavior sequence is too low-entropy to be meaningful.

    Used to filter out degenerate policies (e.g., "always do nothing")
    that achieve high SPC trivially.

    Args:
        actions: Sequence of action indices.
        min_entropy: Minimum behavioral entropy in bits.

    Returns:
        True if the behavior is too low-entropy (should be flagged), False otherwise.
    """
    return behavioral_entropy(actions) < min_entropy


__all__ = [
    "shannon_entropy",
    "distribution_entropy",
    "behavioral_entropy",
    "environmental_entropy",
    "low_information_filter",
]
